File: demo/js/features.js
// BCTS board features on bitboards (PLAN.md §2, features 3-8).
//
// A board is 20 rows of 10-bit ints (bit c set = column c filled), row 0 top,
// row 19 bottom. `boardFeatures` returns the 6 board-only BCTS features of a
// single post-clear board; the placement-dependent landing_height (1) and
// eroded_piece_cells (2) are produced by the engine. This is a faithful port of
// the pure-Python `board_features` reference — the numeric results are integers,
// so JS and Python agree exactly.

export const WIDTH = 10;
export const HEIGHT = 20;
export const FULL_ROW = (1 << WIDTH) - 1; // 0x3FF

function popcount(x) {
  x = x - ((x >> 1) & 0x55555555);
  x = (x & 0x33333333) + ((x >> 2) & 0x33333333);
  x = (x + (x >> 4)) & 0x0f0f0f0f;
  return (Math.imul(x, 0x01010101) >> 24) & 0xff;
}

function rowTransitions(v) {
  // Sequence = [wall, b0..b9, wall]; count adjacent differences. Packed as
  // bit0=left wall, bits1..10=cells, bit11=right wall.
  const a = ((v << 1) | 0x801) & 0xfff;
  return popcount((a ^ (a >> 1)) & 0x7ff);
}

export function boardFeatures(rows) {
  if (rows.length !== HEIGHT) {
    throw new Error(`expected ${HEIGHT} rows, got ${rows.length}`);
  }

  // (3) row transitions.
  let rowTrans = 0;
  for (let r = 0; r < HEIGHT; r++) rowTrans += rowTransitions(rows[r]);

  // (4) column transitions: top border empty, floor filled.
  let colTrans = popcount(rows[0] & FULL_ROW);
  for (let r = 0; r < HEIGHT - 1; r++) {
    colTrans += popcount((rows[r] ^ rows[r + 1]) & FULL_ROW);
  }
  colTrans += popcount((rows[HEIGHT - 1] ^ FULL_ROW) & FULL_ROW);

  // (5)(7)(8): single top-down scan maintaining per-column filled counts.
  let holes = 0;
  let holeDepth = 0;
  let rowsWithHoles = 0;
  const colFilled = new Array(WIDTH).fill(0);
  let aboveOr = 0;
  for (let r = 0; r < HEIGHT; r++) {
    const row = rows[r];
    let anyHole = false;
    for (let c = 0; c < WIDTH; c++) {
      const bit = 1 << c;
      if (aboveOr & bit && !(row & bit)) {
        // Empty now, filled somewhere above => a hole.
        holes++;
        // hole_depth counts ALL filled cells above each hole in its column
        // (not just the contiguous run) — pinned by tests/test_features.py.
        holeDepth += colFilled[c];
        anyHole = true;
      }
      if (row & bit) colFilled[c]++;
    }
    if (anyHole) rowsWithHoles++;
    aboveOr |= row;
  }

  // (6) cumulative wells: empty cell whose left & right neighbors are filled
  // (walls filled); triangular sum over maximal vertical runs.
  let cumWells = 0;
  const runLen = new Array(WIDTH).fill(0);
  for (let r = 0; r < HEIGHT; r++) {
    const row = rows[r];
    const left = ((row << 1) | 1) & FULL_ROW;
    const right = ((row >> 1) | (1 << (WIDTH - 1))) & FULL_ROW;
    const well = ~row & left & right & FULL_ROW;
    for (let c = 0; c < WIDTH; c++) {
      if (well & (1 << c)) {
        runLen[c]++;
        cumWells += runLen[c];
      } else {
        runLen[c] = 0;
      }
    }
  }

  return [rowTrans, colTrans, holes, cumWells, holeDepth, rowsWithHoles];
}

File: tetris/features.py
"""BCTS board features on bitboards (PLAN.md §2, features 3-8).

Six of the eight BCTS features depend only on the post-clear board:
row_transitions, column_transitions, holes, cumulative_wells, hole_depth,
rows_with_holes. The remaining two (landing_height, eroded_piece_cells) depend
on the placement itself and are produced by the engine.

A board is 20 rows of 10-bit ints (bit c set = column c filled), row 0 top,
row 19 bottom. Everything here is bit-parallel: `board_features_batch` scores
many candidate boards at once with numpy for the hot decision path;
`board_features` is a pure-Python reference for a single board.
"""

from __future__ import annotations

import numpy as np

WIDTH = 10
HEIGHT = 20
FULL_ROW = (1 << WIDTH) - 1  # 0x3FF

# --- Lookup tables over all 10-bit row values -------------------------------

_TABLE_SIZE = 1 << WIDTH  # 1024

# Horizontal transitions per row with both side walls counted as filled
# (PLAN.md §2 feature 3). Sequence = [wall, b0..b9, wall]; count adjacent
# differences. Packed as bit0=left wall, bits1..10=cells, bit11=right wall.
_ROW_TRANS = np.empty(_TABLE_SIZE, dtype=np.int64)
_POPCOUNT = np.empty(_TABLE_SIZE, dtype=np.int64)
for _v in range(_TABLE_SIZE):
    _a = (_v << 1) | 0x801
    _ROW_TRANS[_v] = ((_a ^ (_a >> 1)) & 0x7FF).bit_count()
    _POPCOUNT[_v] = _v.bit_count()

_BITPOS = np.arange(WIDTH, dtype=np.uint16)


def _bits(x: np.ndarray) -> np.ndarray:
    """(P,) row ints -> (P, WIDTH) array of 0/1 per column."""
    return ((x[:, None] >> _BITPOS) & 1).astype(np.int64)


def board_features_batch(rows: np.ndarray) -> np.ndarray:
    """Compute the 6 board-only BCTS features for many boards at once.

    Args:
        rows: (P, 20) array of 10-bit row ints (uint16-compatible).

    Returns:
        (P, 6) int64 array:
        [row_transitions, column_transitions, holes, cumulative_wells,
         hole_depth, rows_with_holes].
    """
    rows = np.asarray(rows, dtype=np.uint16)
    if rows.ndim != 2 or rows.shape[1] != HEIGHT:
        raise ValueError(f"expected (P, {HEIGHT}) rows, got {rows.shape}")
    P = rows.shape[0]

    # (3) row transitions: table lookup summed over all 20 rows.
    row_trans = _ROW_TRANS[rows].sum(axis=1)

    # (4) column transitions: top border empty, floor filled.
    col_trans = _POPCOUNT[rows[:, 0]].copy()
    for r in range(HEIGHT - 1):
        col_trans += _POPCOUNT[rows[:, r] ^ rows[:, r + 1]]
    col_trans += _POPCOUNT[rows[:, HEIGHT - 1] ^ FULL_ROW]

    # (5)(6)(7)(8): single top-down scan maintaining per-column state.
    above_or = np.zeros(P, dtype=np.uint16)
    col_filled = np.zeros((P, WIDTH), dtype=np.int64)
    run_len = np.zeros((P, WIDTH), dtype=np.int64)
    holes = np.zeros(P, dtype=np.int64)
    hole_depth = np.zeros(P, dtype=np.int64)
    rows_with_holes = np.zeros(P, dtype=np.int64)
    cum_wells = np.zeros(P, dtype=np.int64)

    for r in range(HEIGHT):
        row = rows[:, r]
        rb = _bits(row)
        ab = _bits(above_or)
        hole_cells = ab & (1 - rb)  # empty now, filled somewhere above
        n_holes = hole_cells.sum(axis=1)
        holes += n_holes
        rows_with_holes += (n_holes > 0).astype(np.int64)
        # hole_depth counts ALL filled cells above each hole in its column (not
        # just the contiguous run) — pinned by test, mirrored in JS engine.
        hole_depth += (hole_cells * col_filled).sum(axis=1)
        col_filled += rb
        above_or = above_or | row

        # (6) cumulative wells: empty cell whose left & right neighbors are
        # filled (walls filled); triangular sum over maximal vertical runs.
        left = ((row << np.uint16(1)) | np.uint16(1)) & np.uint16(FULL_ROW)
        right = ((row >> np.uint16(1)) | np.uint16(1 << (WIDTH - 1))) & np.uint16(FULL_ROW)
        well = (~row) & left & right & np.uint16(FULL_ROW)
        wb = _bits(well)
        run_len = np.where(wb == 1, run_len + 1, 0)
        cum_wells += run_len.sum(axis=1)

    return np.stack(
        [row_trans, col_trans, holes, cum_wells, hole_depth, rows_with_holes],
        axis=1,
    )


def board_features(rows) -> tuple[int, int, int, int, int, int]:
    """Pure-Python reference for the 6 board features of a single board.

    Returns (row_transitions, column_transitions, holes, cumulative_wells,
    hole_depth, rows_with_holes).
    """
    rows = list(rows)
    if len(rows) != HEIGHT:
        raise ValueError(f"expected {HEIGHT} rows, got {len(rows)}")

    row_trans = 0
    for v in rows:
        a = (v << 1) | 0x801
        row_trans += ((a ^ (a >> 1)) & 0x7FF).bit_count()

    col_trans = (rows[0] & FULL_ROW).bit_count()
    for r in range(HEIGHT - 1):
        col_trans += ((rows[r] ^ rows[r + 1]) & FULL_ROW).bit_count()
    col_trans += ((rows[HEIGHT - 1] ^ FULL_ROW) & FULL_ROW).bit_count()

    holes = 0
    hole_depth = 0
    rows_with_holes = 0
    col_filled = [0] * WIDTH
    above_or = 0
    for r in range(HEIGHT):
        row = rows[r]
        hole_bits = above_or & ~row & FULL_ROW
        if hole_bits:
            rows_with_holes += 1
            m = hole_bits
            while m:
                c = (m & -m).bit_length() - 1
                holes += 1
                # hole_depth counts ALL filled cells above each hole in its
                # column (not just the contiguous run) — pinned by test,
                # mirrored in JS engine.
                hole_depth += col_filled[c]
                m &= m - 1
        m = row & FULL_ROW
        while m:
            c = (m & -m).bit_length() - 1
            col_filled[c] += 1
            m &= m - 1
        above_or |= row

    cum_wells = 0
    run_len = [0] * WIDTH
    for r in range(HEIGHT):
        row = rows[r]
        left = ((row << 1) | 1) & FULL_ROW
        right = ((row >> 1) | (1 << (WIDTH - 1))) & FULL_ROW
        well = ~row & left & right & FULL_ROW
        for c in range(WIDTH):
            if well & (1 << c):
                run_len[c] += 1
                cum_wells += run_len[c]
            else:
                run_len[c] = 0

    return (row_trans, col_trans, holes, cum_wells, hole_depth, rows_with_holes)
